Data inspection and cleaning is a crucial phase in any data science or machine learning project, as the quality of your insights heavily depends on the quality of your data. We'll be using a laptop price dataset to walk through common issues you might encounter and how to address them.

In [1]:
import pandas as pd

# Reading from CSV

The most common file format you’ll encounter is CSV (Comma-Separated Values). Loading a CSV file with Pandas is simple:

In [2]:
# df = pd.read_csv("../data/penguins.csv")
df = pd.read_csv("../data/penguins.csv")

In [3]:
df.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female


# Data Inspection
Once we’ve loaded the data, the next step is to get to know it. These quick inspections help identify missing values, strange formats, or inconsistencies.

### First Look at the Data

* df.head() – shows the first 5 rows

* df.tail() – shows the last 5 rows

* df.info() – gives a summary of column names, data types, and non-null counts

* df.describe() – gives descriptive statistics for numerical columns (mean, std, min, max, etc.)

In [4]:
df.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [6]:
df.describe()

,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


These functions help answer key questions:

* What kind of data are we working with?

* Are there missing values?

* Which columns are numerical or categorical?

We can check the dimension of our data by using `df.shape` or `len(df)`.

* `df.shape` – returns a tuple (rows, columns)

* `len(df)` – returns the number of rows

In [7]:
print(df.shape)
print(len(df))

(344, 7)
344


# Data Cleaning
Raw data is rarely perfect. It often contains missing values, wrong data types, duplicate rows, or inconsistent formats. In this lesson, we’ll explore essential techniques to clean and transform your data using Pandas.

Data cleaning is the process of identifying, correcting, or removing incorrect, corrupted, incomplete, duplicate, or irrelevant data within a dataset to improve its quality and usability. This involves **fixing errors such as typos, inconsistencies, missing values, outliers, and formatting issues** to ensure the data is accurate, consistent, and reliable for analysis or machine learning. The goal is to produce a clean dataset that enables more accurate, efficient, and trustworthy decision-making and model training.

# Handling Missing Values
We can handle missing values by dropping them or filling them with an automated workflow. But first, we need to load and inspect our data.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


## How many values are missing in our data?

A quick look at the output of `df.head` and `df.info()` shows that there are some missing values in the data set.

### Checking the number of missing values

In [9]:
df.isna().sum() # or df.isnull()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

In [10]:
df.isna().sum().sum() # or df.isnull().sum()

np.int64(19)

In [11]:
# Sorting the columns based on number of missing values

df.isna().sum().sort_values(ascending=False)

sex                  11
bill_depth_mm         2
bill_length_mm        2
flipper_length_mm     2
body_mass_g           2
island                0
species               0
dtype: int64

### Checking the percentage of missing values

In [12]:
columns_missing = df.isna().sum()
n_rows = len(df)

columns_missing_percentage = columns_missing/n_rows * 100 # or df.isna().mean()*100
columns_missing_percentage

species              0.000000
island               0.000000
bill_length_mm       0.581395
bill_depth_mm        0.581395
flipper_length_mm    0.581395
body_mass_g          0.581395
sex                  3.197674
dtype: float64

In [13]:
# Getting the percentage of the numerical column from the whole data set
# col = df.select_dtypes(include="number").columns
# column_perc = len(col)
# all_df = len(df)
# overall_percent = column_perc/all_df * 100
# overall_percent

In [14]:
# Sort percentage of missing values from highest to lowest

columns_missing_percentage.sort_values(ascending=False)

sex                  3.197674
bill_depth_mm        0.581395
bill_length_mm       0.581395
flipper_length_mm    0.581395
body_mass_g          0.581395
island               0.000000
species              0.000000
dtype: float64

In [15]:
df.shape

(344, 7)

In [16]:
# Total percentage of missing values across rows and columns

total_values = df.shape[0] * df.shape[1]
total_missing = columns_missing.sum()

print(total_values, total_missing, f"{round(total_missing/total_values * 100, 2)}%", sep="\n\n")

2408

19

0.79%


### Exercise 1
Use the "cybersecurity_intrusion_data.csv" file to anwer the following questions.

1. Load the data into a variable named `cs_df`.

2. How many columns and rows are in the dataset?

3. How many cells are missing in the dataset?

4. How many values are missing in each columns in the dataset?

5. What are the proportion of missing missing values in each columns?

#### Answers
### Q1

In [17]:
cs_df = pd.read_csv("../data/cybersecurity_intrusion_data.csv")
cs_df.head()

,session_id,network_packet_size,protocol_type,login_attempts,session_duration,encryption_used,ip_reputation_score,failed_logins,browser_type,unusual_time_access,attack_detected
0,SID_00001,599,TCP,4,492.983263,DES,0.606818,1,Edge,0,1
1,SID_00002,472,TCP,3,1557.996461,DES,0.301569,0,Firefox,0,0
2,SID_00003,629,TCP,3,75.044262,DES,0.739164,2,Chrome,0,1
3,SID_00004,804,UDP,4,601.248835,DES,0.123267,0,Unknown,0,1
4,SID_00005,453,TCP,5,532.540888,AES,0.054874,1,Firefox,0,0


### Q2

In [18]:
cs_df.shape

(9537, 11)

#### the cybersecurity_intrusion_data has 9537 rows and 11 columns
### Q3

In [19]:
cs_df.isna().sum().sum()

np.int64(1966)

#### 1966 cells are misiing in the data set
### Q 4

In [20]:
cs_df.isna().sum()

session_id                0
network_packet_size       0
protocol_type             0
login_attempts            0
session_duration          0
encryption_used        1966
ip_reputation_score       0
failed_logins             0
browser_type              0
unusual_time_access       0
attack_detected           0
dtype: int64

#### Checking the proportion of the missing values in each columns
### Q 5

In [21]:
cs_df.isna().mean()*100

session_id              0.000000
network_packet_size     0.000000
protocol_type           0.000000
login_attempts          0.000000
session_duration        0.000000
encryption_used        20.614449
ip_reputation_score     0.000000
failed_logins           0.000000
browser_type            0.000000
unusual_time_access     0.000000
attack_detected         0.000000
dtype: float64

## Dropping missing values

One of the options we have when dealing with missing values is to just remove any rows or columns that contain missing values. This not recommended for very import projects, so it's better understand why data is missing before deciding whether to remove them or not.

If we are sure you want to drop rows with missing values, pandas has a handy function, `dropna()`.

### Removing rows with missing values

In [22]:
df.dropna() # Drop all rows with missing values

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female
5,Adelie,Torgersen,39.3,20.6,190.0,3650.0,Male
...,...,...,...,...,...,...,...
338,Gentoo,Biscoe,47.2,13.7,214.0,4925.0,Female
340,Gentoo,Biscoe,46.8,14.3,215.0,4850.0,Female
341,Gentoo,Biscoe,50.4,15.7,222.0,5750.0,Male
342,Gentoo,Biscoe,45.2,14.8,212.0,5200.0,Female


### Removing columns with missing values

In [23]:
df_without_na = df.dropna(axis=1) # Drop all columns with missing values
df_without_na

,species,island
0,Adelie,Torgersen
1,Adelie,Torgersen
2,Adelie,Torgersen
3,Adelie,Torgersen
4,Adelie,Torgersen
...,...,...
339,Gentoo,Biscoe
340,Gentoo,Biscoe
341,Gentoo,Biscoe
342,Gentoo,Biscoe


Pandas can also remove columns with missing values based on the number of rows that is missing in that column.

In [24]:
3.197674/100

0.031976740000000003

In [25]:
len(df) * 0.03

10.32

In [26]:
df.isna().sum()/len(df) < 0.03

species               True
island                True
bill_length_mm        True
bill_depth_mm         True
flipper_length_mm     True
body_mass_g           True
sex                  False
dtype: bool

In [27]:
df.loc[:, (df.isna().sum()/len(df) < 0.03)]

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0
3,Adelie,Torgersen,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0
...,...,...,...,...,...,...
339,Gentoo,Biscoe,NaN,NaN,NaN,NaN
340,Gentoo,Biscoe,46.8,14.3,215.0,4850.0
341,Gentoo,Biscoe,50.4,15.7,222.0,5750.0
342,Gentoo,Biscoe,45.2,14.8,212.0,5200.0


In [28]:
df_with_col_3p_na_dropped = df.dropna(thresh=len(df) - int(len(df) * 0.03), axis=1) # Drop rows with at least 3% missing values
df_with_col_3p_na_dropped.shape

(344, 6)

In [29]:
df_with_col_3p_na_dropped.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0
3,Adelie,Torgersen,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0


In [30]:
df.shape

(344, 7)

## Filling missing values

Another option is to try and fill in the missing values. Depending on the type of the data and the distribution of the values, we can replace missing values with zero, mean, median, or mode.

In [31]:
df_with_col_3p_na_dropped.isna().sum().sort_values(ascending=False)/len(df)

bill_depth_mm        0.005814
bill_length_mm       0.005814
flipper_length_mm    0.005814
body_mass_g          0.005814
island               0.000000
species              0.000000
dtype: float64

### Filling missing value in a single column

In [32]:
df_with_col_3p_na_dropped["bill_depth_mm"].isna().sum()

np.int64(2)

In [33]:
bill_depth_mean_filler = df_with_col_3p_na_dropped["bill_depth_mm"].mean()
bill_depth_mean_filler

np.float64(17.151169590643278)

In [34]:
bill_depth_filled = df_with_col_3p_na_dropped["bill_depth_mm"].fillna(bill_depth_mean_filler)
bill_depth_filled.isna().sum()

np.int64(0)

In [35]:
df_with_col_3p_na_dropped["bill_depth_mm"][df_with_col_3p_na_dropped["bill_depth_mm"].isna()]

3     NaN
339   NaN
Name: bill_depth_mm, dtype: float64

In [36]:
bill_depth_filled.loc[[3, 339]]

3      17.15117
339    17.15117
Name: bill_depth_mm, dtype: float64

Now that the missing value if filled, we need to update the dataframe

In [37]:
df_with_col_3p_na_dropped.loc[:, "bill_depth_mm"] = bill_depth_filled

In [38]:
# We can fill missing values in categorical columns by passing the mode (most frequent value to the fillna method)
df_with_col_3p_na_dropped["island"].mode()[0]

'Biscoe'

### Filling missing values in multiple columns

In [39]:
df_with_col_3p_na_dropped.isna().sum()

species              0
island               0
bill_length_mm       2
bill_depth_mm        0
flipper_length_mm    2
body_mass_g          2
dtype: int64

In [40]:
numeric_cols = ["bill_length_mm", "flipper_length_mm", "body_mass_g"]

In [41]:
mult_col_mean_filler = df_with_col_3p_na_dropped[numeric_cols].mean()
mult_col_mean_filler

bill_length_mm         43.921930
flipper_length_mm     200.915205
body_mass_g          4201.754386
dtype: float64

In [42]:
mult_col_filled = df_with_col_3p_na_dropped[numeric_cols].fillna(mult_col_mean_filler)
mult_col_filled.isna().sum()

bill_length_mm       0
flipper_length_mm    0
body_mass_g          0
dtype: int64

In [43]:
all([True, True, False, True])

False

In [44]:
df_with_col_3p_na_dropped[df_with_col_3p_na_dropped.isna().any(axis=1)]

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
3,Adelie,Torgersen,NaN,17.15117,NaN,NaN
339,Gentoo,Biscoe,NaN,17.15117,NaN,NaN


In [45]:
mult_col_filled.loc[[3, 339]]

,bill_length_mm,flipper_length_mm,body_mass_g
3,43.92193,200.915205,4201.754386
339,43.92193,200.915205,4201.754386


# Handling Duplicate values

We can inspect duplicate rows in our data by using our DataFrame's `duplicated()`.

In [46]:
df2 = pd.DataFrame(
    {
        "brand": ["Volvo", "Volvo", "BMW", "BMW", "BMW", "BMW"],
        "color": ["red", "red", "blue", "green", "green", "green"],
        "top speed": [250, 250, 300, 310, 280, 310],
    }
)
df2

,brand,color,top speed
0,Volvo,red,250
1,Volvo,red,250
2,BMW,blue,300
3,BMW,green,310
4,BMW,green,280
5,BMW,green,310


In [47]:
df2.duplicated() # Check for duplicate rows

0    False
1     True
2    False
3    False
4    False
5     True
dtype: bool

In [48]:
df2.duplicated().sum() # Check for number of duplicate rows

np.int64(2)

## Removing duplicate values

We can drop duplicate rows by using the DataFrame's `drop_duplicates` method.

In [49]:
df2_duplicates_dropped = df2.drop_duplicates() # Drop duplicate rows
df2_duplicates_dropped

,brand,color,top speed
0,Volvo,red,250
2,BMW,blue,300
3,BMW,green,310
4,BMW,green,280


In [50]:
df2_duplicates_dropped.duplicated().sum()

np.int64(0)

# Handling Inconsistent Entries

Another thing to take note of our data cleaning stage is inconsistent values in certain columns in our data set.

In [51]:
data = {
    "Name": ["Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Jane", "John"],
    "City": ["New York", "new york", "NYC", "Los Angeles", "LA", "los angeles", " New York", "losangeles"],
    "Gender": ["Male", "male", "FEMALE", "Female", "female", "MALE", "MaLE", " FeMaLE"],
}

df = pd.DataFrame(data)
df

,Name,City,Gender
0,Alice,New York,Male
1,Bob,new york,male
2,Charlie,NYC,FEMALE
3,David,Los Angeles,Female
4,Eve,LA,female
5,Frank,los angeles,MALE
6,Jane,New York,MaLE
7,John,losangeles,FeMaLE


In [52]:
print("Unique values in 'City':", df['City'].unique())
print("Unique values in 'Gender':", df['Gender'].unique())

Unique values in 'City': ['New York' 'new york' 'NYC' 'Los Angeles' 'LA' 'los angeles' ' New York'
 'losangeles']
Unique values in 'Gender': ['Male' 'male' 'FEMALE' 'Female' 'female' 'MALE' 'MaLE' ' FeMaLE']


Looking at the data, we can see there are inconsistent values in the "City" and "Gender" column. For example, in the "City" column "New York" and "new york". There are also some values with trailing white spaces in the column.

The first step in fixing this is to standardise the values and remove any white spaces that might be before or after each values.

In [53]:
df["City"] = df["City"].str.lower()
df["Gender"] = df["Gender"].str.lower()

In [54]:
print("Unique values in 'City':", df['City'].unique())
print("Unique values in 'Gender':", df['Gender'].unique())

Unique values in 'City': ['new york' 'nyc' 'los angeles' 'la' ' new york' 'losangeles']
Unique values in 'Gender': ['male' 'female' ' female']


All values are now standardised to be in lower case. But there are still values with white spaces. We can fix that using strings "strip" method.

In [55]:
df["City"] = df["City"].str.strip()
df["Gender"] = df["Gender"].str.strip()

In [56]:
print("Unique values in 'City':", df['City'].unique())
print("Unique values in 'Gender':", df['Gender'].unique())

Unique values in 'City': ['new york' 'nyc' 'los angeles' 'la' 'losangeles']
Unique values in 'Gender': ['male' 'female']


All inconisistent entries in the gender column has been fixed. However, we still need to do one more step to completely fix the values in the "City" column. For example, "new york" and "nyc". We can fix this by mapping similar value to a consistent label.

In [57]:
city_mapping = {"nyc": "new york", "la": "los angeles", "losangeles": "los angeles"}
df["City"] = df["City"].replace(city_mapping)

In [58]:
print("Unique values in 'City':", df['City'].unique())
print("Unique values in 'Gender':", df['Gender'].unique())

Unique values in 'City': ['new york' 'los angeles']
Unique values in 'Gender': ['male' 'female']


In [59]:
df

,Name,City,Gender
0,Alice,new york,male
1,Bob,new york,male
2,Charlie,new york,female
3,David,los angeles,female
4,Eve,los angeles,female
5,Frank,los angeles,male
6,Jane,new york,male
7,John,los angeles,female


### Exercise 2

Use the data given below to answer the following questions:

```python
data = {
    "UserID": [101, 102, 103, 104, 105, 106, 107],
    "Country": ["usa", "USA", "United States", "u.s.", "India", "INDIA", "inDia"],
    "Device": ["iPhone", "iphone", "Android", "android ", "iOS", "IOS", "ANDROID"]
}

df = pd.DataFrame(data)
```

1. Inspect the unique values in the categorical columns.

2. Convert the values to lower case and remove any white spaces included.

3. Map inconsistent values to standard labels.

## Answers

In [60]:
data = {
    "UserID": [101, 102, 103, 104, 105, 106, 107],
    "Country": ["usa", "USA", "United States", "u.s.", "India", "INDIA", "inDia"],
    "Device": ["iPhone", "iphone", "Android", "android ", "iOS", "IOS", "ANDROID"]
}

df3 = pd.DataFrame(data)
df3

,UserID,Country,Device
0,101,usa,iPhone
1,102,USA,iphone
2,103,United States,Android
3,104,u.s.,android
4,105,India,iOS
5,106,INDIA,IOS
6,107,inDia,ANDROID


In [61]:
print ("Unique values in the country's column: ", df3["Country"].unique())
print ("Unique values in the country's column: ", df3["Device"].unique())

Unique values in the country's column:  ['usa' 'USA' 'United States' 'u.s.' 'India' 'INDIA' 'inDia']
Unique values in the country's column:  ['iPhone' 'iphone' 'Android' 'android ' 'iOS' 'IOS' 'ANDROID']


In [62]:
df3["Country"] = df3["Country"].str.lower()
df3["Device"] = df3["Device"].str.lower()
print (df3["Country"])
df3["Device"]

0              usa
1              usa
2    united states
3             u.s.
4            india
5            india
6            india
Name: Country, dtype: object


0      iphone
1      iphone
2     android
3    android 
4         ios
5         ios
6     android
Name: Device, dtype: object

In [63]:
df3["Country"] = df3["Country"].str.strip()
df3["Device"] = df3["Device"].str.strip()

In [64]:
Country_mapping = {"u.s.": "usa", "united states": "usa"}
Device_mapping = {"ios": "android"}
df3["Country"] = df3["Country"].replace(Country_mapping)
df3["Device"] = df3["Device"].replace(Device_mapping)

In [65]:
df3["Country"].unique()

array(['usa', 'india'], dtype=object)

In [66]:
df3

,UserID,Country,Device
0,101,usa,iphone
1,102,usa,iphone
2,103,usa,android
3,104,usa,android
4,105,india,android
5,106,india,android
6,107,india,android


# Fixing Wrong Data Types

When working with real world data. Columns may have incorrect data types. For example, numbers stored as string, dates stored as plain text, boolean values reprersented as string.

In [67]:
data = {
    "CustomerID": ["001", "002", "003", "004"],
    "Age": ["25", "30", "22", "27"],  # Should be integers
    "SignupDate": [
        "2024/01/10",
        "2024/02/15",
        "2024/03/05",
        "2024/04/20",
    ],  # Should be datetime
    "IsActive": ["True", "False", "TRUE", "false"],  # Should be boolean
}

df = pd.DataFrame(data)
df


,CustomerID,Age,SignupDate,IsActive
0,001,25,2024/01/10,True
1,002,30,2024/02/15,False
2,003,22,2024/03/05,TRUE
3,004,27,2024/04/20,false


In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   CustomerID  4 non-null      object
 1   Age         4 non-null      object
 2   SignupDate  4 non-null      object
 3   IsActive    4 non-null      object
dtypes: object(4)
memory usage: 260.0+ bytes


In [69]:
df.dtypes

CustomerID    object
Age           object
SignupDate    object
IsActive      object
dtype: object

In [70]:
# Convert Age to integer
df["Age"] = df["Age"].astype(int)

In [71]:
# Convert IsActive to boolean
df["IsActive"] = df["IsActive"].str.lower().map({"true": True, "false": False})

In [72]:
# Convert SignupDate to datetime
df["SignupDate"] = pd.to_datetime(df["SignupDate"])

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   CustomerID  4 non-null      object        
 1   Age         4 non-null      int64         
 2   SignupDate  4 non-null      datetime64[ns]
 3   IsActive    4 non-null      bool          
dtypes: bool(1), datetime64[ns](1), int64(1), object(1)
memory usage: 232.0+ bytes


# Parsing Dates

Pandas also allows us to specify the format for parsing dates by providing a string representation of how the date is strucktured. By providing a format string containing special format codes (like %Y for year, %m for month, %d for day), we can customize how the date and time information is represented. For example, "2024.05.08" has the format "%Y.%m.%d"

In [74]:
date_data = {
    "OrderID": [1, 2, 3, 4],
    "OrderDate": ["2024,01,10", "2024,02,15", "2024,03,05", "2024,01,04"]
}

df_dates = pd.DataFrame(date_data)
df_dates

,OrderID,OrderDate
0,1,"2024,01,10"
1,2,"2024,02,15"
2,3,"2024,03,05"
3,4,"2024,01,04"


In [75]:
pd.to_datetime(df_dates["OrderDate"], format="%Y,%m,%d")

0   2024-01-10
1   2024-02-15
2   2024-03-05
3   2024-01-04
Name: OrderDate, dtype: datetime64[ns]

In [76]:
df_dates["OrderDateParsed"] = pd.to_datetime(df_dates["OrderDate"], format="%Y,%m,%d")  # Specify format
df_dates

,OrderID,OrderDate,OrderDateParsed
0,1,"2024,01,10",2024-01-10
1,2,"2024,02,15",2024-02-15
2,3,"2024,03,05",2024-03-05
3,4,"2024,01,04",2024-01-04


There are other possible [parts](https://strftime.org/) we can specify providing the formats for our date string but the most common but the most common are %d for day, %m for month, %y for a two-digit year and %Y for a four digit year.

Dataset sometimes contain different date formats:
- `"2024-01-10"` (ISO format)

- `"2024/02/15"` (slashes)

- `"March 5, 2024"` (long format)

- `"04-01-2024"` (ambiguous: is it April 1 or Jan 4?)

In [77]:
date_data = {
    "OrderID": [1, 2, 3, 4],
    "OrderDate": ["2024-01-10", "2024/02/15", "March 5, 2024", "04-01-2024"]  # Different formats
}

df_dates = pd.DataFrame(date_data)
df_dates

,OrderID,OrderDate
0,1,2024-01-10
1,2,2024/02/15
2,3,"March 5, 2024"
3,4,04-01-2024


In [78]:
# Let pandas automatically infer formats using, assuming January 4 for 04-01-2024
df_dates["OrderDate_parsed"] = pd.to_datetime(df_dates["OrderDate"], format="mixed")

df_dates

,OrderID,OrderDate,OrderDate_parsed
0,1,2024-01-10,2024-01-10
1,2,2024/02/15,2024-02-15
2,3,"March 5, 2024",2024-03-05
3,4,04-01-2024,2024-04-01


Pandas automatically converts the dates for us. However, it couldn't the format for "04-01-2024" as January 4.

In [79]:
df_dates["OrderDate_parsed_day_first"] = pd.to_datetime(df_dates["OrderDate"], format="mixed", dayfirst=True)

df_dates

,OrderID,OrderDate,OrderDate_parsed,OrderDate_parsed_day_first
0,1,2024-01-10,2024-01-10,2024-01-10
1,2,2024/02/15,2024-02-15,2024-02-15
2,3,"March 5, 2024",2024-03-05,2024-03-05
3,4,04-01-2024,2024-04-01,2024-01-04


If we don't specify a format. Pandas will raise an error for values that does not follow the default format. We can silence this error and replace bad values with missing values by passing `error="coerce"` to the `pd.to_datetime` function.

In [80]:
pd.to_datetime(df_dates["OrderDate"], errors="coerce")

0   2024-01-10
1          NaT
2          NaT
3          NaT
Name: OrderDate, dtype: datetime64[ns]

## Exercise 3

1. Suppose you have the dateset given below:
    ```python
    data = {
        "ProductID": ["1001", "1002", "1003"],
        "Price": ["19.99", "25.00", "15.50"],
        "Stock": ["10", "5", "8"]
    }

    df = pd.DataFrame(data)
    ```
    Convert `ProductID` to int, `Price` and `Stock` to float. And print the column data types.

2. Suppose you have a DataFrame of event dates:
    ```python
    data = {
        "Event": ["A", "B", "C", "D"],
        "Date": ["2024-01-01", "01/15/2024", "Feb 5, 2024", "2024.03.01"]
    }

    df = pd.DataFrame(data)
    ```
    Convert the `Date` column to datetime format and store the result in a new column called `ParsedDate`

### Answers
### Q1

In [81]:
data = {
    "ProductID": ["1001", "1002", "1003"],
    "Price": ["19.99", "25.00", "15.50"],
    "Stock": ["10", "5", "8"]
}
df = pd.DataFrame(data)
df

,ProductID,Price,Stock
0,1001,19.99,10
1,1002,25.00,5
2,1003,15.50,8


In [82]:
df["ProductID"].astype(int)

0    1001
1    1002
2    1003
Name: ProductID, dtype: int64

In [83]:
df[["Price", "Stock"]].astype(float)

,Price,Stock
0,19.99,10.0
1,25.00,5.0
2,15.50,8.0


### Q2

In [84]:
data = {
    "Event": ["A", "B", "C", "D"],
    "Date": ["2024-01-01", "01/15/2024", "Feb 5, 2024", "2024.03.01"]
}

In [95]:
df_data = pd.DataFrame(data)
df_data

,Event,Date
0,A,2024-01-01
1,B,01/15/2024
2,C,"Feb 5, 2024"
3,D,2024.03.01


In [99]:
df["Date_parsed"] = pd.to_datetime(df_data["Date"], dayfirst=True, format="mixed")
df

,Date,Date_parsed
0,2024-01-01 00:00:00,2024-01-01
1,2024-01-15 00:00:00,2024-01-15
2,2024-02-05 00:00:00,2024-02-05
3,2024-03-01 00:00:00,2024-03-01
Date_parsed,"DatetimeIndex(['2024-01-01', '2024-01-15', '20...",NaT
